packages/db-kit/src/cli/decodeTransaction/useDecodedTransaction.ts

```diff
@@ -0,0 +1,70 @@
+import { useState, useEffect } from "react";
+import type { Transaction, TransactionReceipt } from "web3-core";
+import type { WireDecoder, Log } from "@truffle/decoder";
+import type { LogDecoding, CalldataDecoding } from "@truffle/codec";
+
+export interface UseDecodedTransactionOptions {
+  decoder: WireDecoder;
+  transaction: Transaction;
+  receipt: TransactionReceipt;
+}
+
+export interface SummaryState {
+  decoding: CalldataDecoding | undefined;
+  complete: boolean;
+}
+
+export interface EventState {
+  log: Log;
+  decoding: LogDecoding | undefined;
+  complete: boolean;
+}
+
+export function useDecodedTransaction({
+  decoder,
+  transaction,
+  receipt
+}: UseDecodedTransactionOptions): {
+  summary: SummaryState;
+  events: EventState[];
+} {
+  const [summary, setSummaryState] = useState<SummaryState>({
+    decoding: undefined,
+    complete: false
+  });
+
+  const eventStates: {
+    state: EventState;
+    setState: (state: EventState) => void;
+  }[] = [];
+  for (const log of receipt.logs) {
+    const [state, setState] = useState<EventState>({
+      log,
+      decoding: undefined,
+      complete: false
+    });
+    eventStates.push({ state, setState });
+  }
```

Comment:
Interesting, don't see this structure too often.

I feel like a single state var like `const [receiptLogs, setReceiptLogs] = useState()` is more conventional.

Reply:
Then on line 57 you construct the new array and `setState` just once

Reply:
I had it your way and it was giving me a lot of trouble. If I get to it, I'll try switching back to that. How big a deal do you think it is, having all these extra states?

Reply:
I kinda like the `state` and `setState`, personally. Wasn't aware that the other way was the more conventional approach!

Reply:
React Hooks always have to run in the same order. So you're letting something from the outside (the `receipt.logs`) control the ordering that the hooks are run in.